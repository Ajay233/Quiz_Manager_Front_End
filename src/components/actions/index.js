export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
