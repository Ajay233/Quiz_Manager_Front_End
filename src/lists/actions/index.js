import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
