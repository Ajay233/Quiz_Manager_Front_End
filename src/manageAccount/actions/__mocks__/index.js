export const updateUser = jest.fn(() => { return { type: "SET_USER", payload: "" } })
