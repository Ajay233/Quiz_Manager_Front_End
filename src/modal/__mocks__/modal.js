export const Modal = jest.fn(() => { return null })
