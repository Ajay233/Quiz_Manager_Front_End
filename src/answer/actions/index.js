import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
