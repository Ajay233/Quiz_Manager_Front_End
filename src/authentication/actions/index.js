import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
