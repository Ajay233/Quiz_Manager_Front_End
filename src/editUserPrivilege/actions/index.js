import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
