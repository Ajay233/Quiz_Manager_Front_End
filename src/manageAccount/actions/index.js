import { put } from '../../axiosRequests/requests'
import { setUserData } from '../../authentication/actions'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const updateUser = (data, jwt) => {
  const emailChangeMsg = "Profile information updated - Your email was changed so your access has been restriced until your email is verified"
  const errorMsg = "Error updating your profile data, please check the details you provided and try again";
  const emailTaken = "An account with that email already exists"
  return (dispatch) => {
    return put('users/update', data, jwt).then((response) => {
      if(response.data.jwt === undefined){
        dispatch(setUserData(response.data, jwt))
        dispatch(setNotification("Profile information updated", "success", true))
      } else {
        dispatch(setUserData(response.data.user, response.data.jwt))
        dispatch(setNotification(emailChangeMsg, "success", true))
      }
    }).catch((error) => {
      console.log(error)
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.data === emailTaken){
        dispatch(setNotification(emailTaken, "error", true))
      } else {
        dispatch(setNotification(errorMsg, "error", true))
      }
    });
  }
}
