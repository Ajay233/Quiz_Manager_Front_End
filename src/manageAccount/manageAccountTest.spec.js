import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import Modal from '../modal/modal'
import { ManageAccount, mapStateToProps } from './ManageAccount'
import configureStore from 'redux-mock-store'
import mockAxios from 'jest-mock-axios'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import history from '../history'

import { setNotification } from '../notifications/actions'
import { logOut } from '../authentication/actions'
import { showModal, hideModal } from '../modal/actions'
import { sessionExpired } from '../utils/session'

const mockStore = configureStore({})

jest.mock("../axiosRequests/axiosUtil")
jest.mock("../notifications/actions")
jest.mock("../authentication/actions")
jest.mock("../utils/session")
jest.mock("../modal/actions")
jest.mock("../history")

describe("mapStateToProps", () => {
  it("should map state to props", () => {
    const appState = {
      userData: {
        forename: "testForename",
        surname: "testSurname",
        email: "testEmail"
      },
      notificationData:{
        message: "",
        type: "",
        show: false,
        timed: true
      },
      modalState: {
        showModal: false,
        showModal2: false,
        showModal3: false
      }
    }

    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("Manage Account", () => {

  let modalState;
  let userData;
  let notificationData;
  let store;

  beforeEach(() => {
    modalState = {
      showModal: false,
      showModal2: false,
      showModal3: false
    }

    userData = {
      forename: "testForename",
      surname: "testSurname",
      email: "testEmail"
    }

    notificationData = {
      message: "",
      type: "",
      show: false,
      timed: true
    }

    store = mockStore({
      userData: {
        forename: "testForename",
        surname: "testSurname",
        email: "testEmail"
      },
      notificationData:{
        message: "",
        type: "",
        show: false,
        timed: true
      }
    })

    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    })
  })

  afterEach(() => {
    setNotification.mockClear()
    mockAxios.reset()
  })

  it("should call showModal when delete account is clicked", () => {

    const component = renderer.create(
      <Provider store={store}>
        <ManageAccount
          modalState={modalState}
          userData={userData}
          notificationData={notificationData}
        />
      </Provider>
    );

    expect(component).toMatchSnapshot()
  })

  it("should call logOut, history.push and setNotification when an account is deleted", () => {
    const component = render(
      <Provider store={store}>
        <ManageAccount
          modalState={modalState}
          userData={userData}
          notificationData={notificationData}
          showModal={showModal}
        />
      </Provider>
    );

    fireEvent.click(component.getByTestId("delete-account"))
    expect(showModal).toHaveBeenCalledTimes(1)
  })

  it("should call logOut, history.push and setNotification when an account is deleted", () => {
    modalState = {
      showModal: true,
      showModal2: false,
      showModal3: false
    }

    const component = render(
      <Provider store={store}>
        <Router history={history}>
          <ManageAccount
            modalState={modalState}
            userData={userData}
            notificationData={notificationData}
            showModal={showModal}
            logOut={logOut}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    );

    const requestResponse = {
      data: "DELETED"
    }

    fireEvent.click(component.getByTestId("modal-delete-button"))
    mockAxios.mockResponse(requestResponse)
    expect(logOut).toHaveBeenCalledTimes(1)
    expect(history.push).toHaveBeenCalledTimes(1)
    expect(setNotification).toHaveBeenCalledTimes(1)
  })

  it("should call sessionExpired when a 403 error status is recieved", () => {
    modalState = {
      showModal: true,
      showModal2: false,
      showModal3: false
    }

    const component = render(
      <Provider store={store}>
        <Router history={history}>
          <ManageAccount
            modalState={modalState}
            userData={userData}
            notificationData={notificationData}
            showModal={showModal}
            logOut={logOut}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    );

    const errorResponse = {
      response: {
        status: 403
      }
    }

    fireEvent.click(component.getByTestId("modal-delete-button"))
    mockAxios.mockError(errorResponse)
    expect(sessionExpired).toHaveBeenCalledTimes(1)
  })

  it("should call setNotification for any other error status", () => {
    modalState = {
      showModal: true,
      showModal2: false,
      showModal3: false
    }

    const component = render(
      <Provider store={store}>
        <Router history={history}>
          <ManageAccount
            modalState={modalState}
            userData={userData}
            notificationData={notificationData}
            showModal={showModal}
            logOut={logOut}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    );

    const errorResponse = {
      response: {
        status: 404
      }
    }

    fireEvent.click(component.getByTestId("modal-delete-button"))
    mockAxios.mockError(errorResponse)
    expect(setNotification).toHaveBeenCalledTimes(1)
  })
})
