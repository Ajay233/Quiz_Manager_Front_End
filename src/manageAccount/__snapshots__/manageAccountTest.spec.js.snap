// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Account should call showModal when delete account is clicked 1`] = `
<div
  id="editProfileContainer"
>
  <div
    className="hide"
  >
    <div
      className="modalContent"
      onClick={[Function]}
    >
      <span
        className="close"
      >
        <i
          className="fas fa-times-circle red"
        />
      </span>
      <div
        className="modalHeader"
      >
        Account
      </div>
      <div
        className="modalMessage"
      >
        <img
          alt=""
          className="modalImg"
          src="warning.png"
        />
        <div
          className="txt"
        >
          You are about to delete your account.  Once deleted your account will no longer be retreivable
        </div>
      </div>
      <div
        className="modalActions"
      >
        <button
          className="delete"
          data-testid="modal-delete-button"
          onClick={[Function]}
        >
          Delete
        </button>
        <button
          className="submit"
          data-testid="modal-cancel-button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
  <div
    id="editProfileTitle"
  >
    <div
      className="titleImgContainer"
    >
      <img
        alt=""
        id="idImg"
        src="id.png"
      />
    </div>
    <div
      className="titleTextContainer"
    >
      <div
        className="titleText"
      >
        My Account
      </div>
    </div>
  </div>
  <div
    id="editProfileFormContainer"
  >
    <div
      id="editUserDetails"
    >
      <form
        className="profileForm"
        onSubmit={[Function]}
      >
        <div>
          <div
            id="editUserDetailsTitle"
          >
            Edit Personal Details
          </div>
        </div>
        <div>
          <label>
            Forename
          </label>
          <input
            className="inputBox"
            name="forename"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="testForename"
            value=""
          />
        </div>
        <div>
          <label>
            Surname
          </label>
          <input
            className="inputBox"
            name="surname"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="testSurname"
            value=""
          />
        </div>
        <div
          className="accountFieldText"
          id="userEmail"
        >
          Your email:
          <span>
             
            testEmail
             
            <span
              className="red rightAlign"
            >
              <b>
                Unverified
              </b>
               
              <i
                className="fas fa-times-circle"
              />
            </span>
          </span>
        </div>
        <div>
          <label>
            New Email
          </label>
          <input
            className="inputBox"
            name="newEmail"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="testEmail"
            value=""
          />
        </div>
        <button
          className="submit"
          id="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      className="formDividers right"
    />
    <div
      className="formDividers left"
    />
    <div
      id="changePassword"
    >
      <form
        className="changePasswordForm"
        onSubmit={[Function]}
      >
        <div>
          <div
            id="changePasswordTitle"
          >
            Change Password
          </div>
        </div>
        <div>
          <label>
            Current Password:
          </label>
          <input
            className="inputBox"
            id="ChangePassword - password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="password"
            value=""
          />
        </div>
        <div>
          <label>
            New Password:
          </label>
          <input
            className="inputBox"
            id="ChangePassword - newPassword"
            name="newPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="password"
            value=""
          />
        </div>
        <div>
          <label>
             Re-type Password:
          </label>
          <input
            className="inputBox"
            id="ChangePassword - retypedPassword"
            name="retypedPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            type="password"
            value=""
          />
        </div>
        <button
          className="submit"
          id="submitChangePassword"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
  <hr />
  <div>
    <div
      id="editUserDetails"
    >
      <form
        className="profileForm"
        onSubmit={[Function]}
      >
        <div>
          <div
            id="editUserDetailsTitle"
          >
            Permission Level
          </div>
        </div>
        <div>
          <div
            className="accountFieldText"
          >
            Current permission level: 
          </div>
        </div>
        <div>
          <label>
            Permission level
          </label>
          <div>
            <select
              className="inputBox select-half"
              name="permission"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onFocus={[Function]}
              value=""
            >
              <option
                disabled={true}
                value=""
              >
                Select a category
              </option>
              <option
                value="READ-ONLY"
              >
                READ-ONLY
              </option>
              <option
                value="ADMIN"
              >
                ADMIN
              </option>
              <option
                value="SUPER-USER"
              >
                SUPER-USER
              </option>
            </select>
          </div>
        </div>
        <button
          className="submit"
          disabled={true}
          id="submitPermissionRequest"
        >
          <i
            className="far fa-envelope"
          />
           Send Request
        </button>
      </form>
    </div>
  </div>
  <hr />
  <div
    id="deleteAccountContainer"
  >
    <div
      id="deleteAccountTitle"
    >
      Delete Account
    </div>
    <p>
      Please note that once you have deleted your account it will not be possible to retreive it.
    </p>
    <button
      className="delete"
      data-testid="delete-account"
    >
      Delete
    </button>
  </div>
</div>
`;
